### PR TITLE
Swap and add encoding

### LINKED ISSUE #4: Add Swap and Add Calldata Formatter

### PR DESCRIPTION
**Pull request by ewilz**
closes #4

### CODE REVIEW

File: src/swapRouter.ts
Comment (ewilz) [inline src/swapRouter.ts:62]: should this live in `SwapOptions` or be its own parameter?

File: src/swapRouter.ts
Comment (ewilz) [inline src/swapRouter.ts:365]: per the object at the top, might just consolidate this into options: SwapOptions

File: src/swapRouter.ts
Comment (ewilz) [inline src/swapRouter.ts:411]: if one of these tokens were native, we are returning WETH9 atm. can change this.

File: src/approveAndCall.ts
Comment (NoahZinsmeister) [inline src/approveAndCall.ts:5]: probably should add `callPositionManager` into this interface

i wonder whether it should accept an array of calls, which would be passed on directly if length == 1, and otherwise formatted against the nft `multicall`? i kinda think latter?

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:62]: i think this should be in a separate interface

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:66]: ah, this is meant to represent the optimization value indicating `address(this)`? let's use `ADDRESS_THIS` from constants for consistency

File: src/swapRouter.ts
Comment (NoahZinsmeister) [reply src/swapRouter.ts:417]: yeah we should always refund in the same currency

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:388]: wow, interesting, never would have thought to use `pay` to wrap ETH into WETH like this, but i guess it works? if we end up doing it this way, deserves a special-purpose function in the v3-sdk `payments.ts` though.

however, i think we should probably add a special-purpose smart contract function for this use case, because the `payer` argument isn't doing anything here, and also we're triggering an extraneous `IWETH9(WETH9).transfer(recipient, value);` call.

wdyt about a `PeripheryPaymentsExtended#wrapETH` or similar?

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:415]: this needs to be wrapped in an `ApproveAndCall#callPositionManager`

**Review by NoahZinsmeister — COMMENTED** [review]
good start! i think i like the general architecture of a public swap-only method and a public swap-and-add method, both of which are backed by the same underlying swap encoding logic

File: src/swapRouter.ts
Comment (ewilz) [reply src/swapRouter.ts:62]: would break this line, but I can make an optional parameter for `encodeSwaps` instead. https://github.com/Uniswap/router-sdk/pull/7/files#diff-d6c50d2e115ee74f2bd01e291289d53d8d555af6622dbc13ad117bd108177792R268

File: src/swapRouter.ts
Comment (ewilz) [reply src/swapRouter.ts:388]: yup, I think we need a wrapETH function!

**Comment by ewilz** [comment]
> good start! i think i like the general architecture of a public swap-only method and a public swap-and-add method, both of which are backed by the same underlying swap encoding logic

thanks, @NoahZinsmeister. definitely open to other ideas on organizing this too if you had any.

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:372]: hmm i think we have to use the chainId from the underlying tokens?

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:376]: hmm, i guess this is for the case when the user already has some of the output currency, but it's not in the appropriate ratio with the input? so you need to either wrap or pull after the swap? guess that makes sense

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:379]: so here i guess we're using native output as a flag for wanting to pay the remainder in ETH, or not for WETH?

File: src/swapRouter.ts
Comment (NoahZinsmeister) [inline src/swapRouter.ts:390]: ah, so this is wasteful, in the vast majority of cases we only need to do this once per token.

perhaps we need a flag for whether approvals are needed?

**Review by NoahZinsmeister — COMMENTED** [review]
left some comments, looking pretty good! i'm guessing you've validated this calldata generation against the contracts in hardhat? which i should look for here? https://github.com/Uniswap/swap-router-contracts/pull/34/files

File: src/swapRouter.ts
Comment (ewilz) [reply src/swapRouter.ts:376]: in this situation, the user is only getting a partial amount of a token from the swap, but already owns the rest.

File: src/swapRouter.ts
Comment (ewilz) [reply src/swapRouter.ts:379]: hmmm....yah. I guess this is a weird situation in which you list your output token as native, even though it will never actually get converted back to native since it's going into the position. But it allows you to send the "remaining" portion as native. If the wording is too indirect I can try and think of something else...lmk what you think.

File: src/swapRouter.ts
Comment (ewilz) [reply src/swapRouter.ts:376]: This is also why "single asset add" always confused me as a term. I always thought of it as being potentially 2 assets as well.

File: src/swapRouter.ts
Comment (NoahZinsmeister) [reply src/swapRouter.ts:379]: maybe just a brief comment explaining? i think it's fine otherwise

**Review by NoahZinsmeister — APPROVED** [review]
LGTM if you're confident!

**Pull request merged**
